### ISSUE TITLE
1.7-rc.6, loads, but crashes on first reload, symlinked packages starting with @

### ISSUE THREAD

**Issue by houshuang** [meteor/meteor#9882]
OSX 10.12.6, Node 8.9.4

Because we are using Yarn to manage our monorepo (of which the Meteor app is one of the packages), and Meteor does not look for npm packages that have been hoisted to a parent directory, we currently symlink all the packages in ../node_modules into the Meteor apps node_modules.

This has worked perfectly for a long time, however since 1.6.2-beta.18 or something like that, Meteor will start on the first run without any issues, but the moment we trigger a reload by editing a file, it fails.

```js
=> Started proxy.
=> Started MongoDB.
I20180512-12:48:27.016(2)? Meteor settings { public: {} }
I20180512-12:48:28.022(2)? Running shareDB server
=> Started your app.

=> App running at: http://localhost:3000/
/Users/stian/.meteor/packages/coffeescript/.1.0.17.ztkmml.8voq++os+web.browser+web.cordova/plugin.compileCoffeescript.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:116
      throw error;
      ^

Error: ENOENT: no such file or directory, unlink 'npm/node_modules/@babel'
    at Object.fs.unlinkSync (fs.js:1061:18)
    at Object.wrapper [as unlink] (/tools/fs/files.js:1597:35)
    at parts.forEach.part (/tools/isobuild/builder.js:138:19)
    at Array.forEach (<anonymous>)
    at Builder._ensureDirectory (/tools/isobuild/builder.js:130:11)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:438:10)
    at optimisticReaddir.forEach.item (/tools/isobuild/builder.js:441:12)
    at Array.forEach (<anonymous>)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:440:35)
    at Builder.copyNodeModulesDirectory (/tools/isobuild/builder.js:410:12)
    at /tools/isobuild/bundler.js:2307:17
    at Function._.each._.forEach (/Users/stian/.meteor/packages/meteor-tool/.1.7.0-rc.6.5p2mbn.25ps++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:2266:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2538:13)
    at /tools/isobuild/bundler.js:2662:30
    at Object.keys.forEach.name (/tools/isobuild/bundler.js:2797:9)
    at Array.forEach (<anonymous>)
    at /tools/isobuild/bundler.js:2787:26
    at /tools/isobuild/bundler.js:3128:22
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2961:31)
    at files.withCache (/tools/isobuild/bundler.js:2908:32)
    at Object.withCache (/tools/fs/files.js:1700:12)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2908:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
```

**Comment by houshuang** [comment]
After scattering console.logs all over the Builder, I've isolated the line where the error is happening. Somehow it only triggers on symlinked node directories which have a @ in them (perhaps because they contain multiple packages, and not a package.json file?). 

By wrapping the single line in a try...catch, Meteor is able to continue working (seemingly) perfectly.

Here is a code snippet from `isobuild/builder.js` (see where I added the try...catch). So I found the error location, and trigger, but after spending some time experimenting I was not able to fix it (other than this try...catch), and I have no idea why it would choke on these specific files.

```js
  _ensureDirectory(relPath) {
    const parts = files.pathNormalize(relPath).split(files.pathSep);

    if (parts.length > 1 && parts[parts.length - 1] === '') {
      // remove trailing slash
      parts.pop();
    }

    const partsSoFar = [];
    parts.forEach(part => {
      partsSoFar.push(part);
      const partial = partsSoFar.join(files.pathSep);

      if (!(partial in this.usedAsFile)) {
        let needToMkdir = true;

        if (partial in this.previousUsedAsFile) {
          if (this.previousUsedAsFile[partial]) {
            // was previously used as file, delete it, create a directory
            try {
              files.unlink(partial);
            } catch (e) {
              console.log(e);
            }
          } else {
            // is already a directory
            needToMkdir = false;
          }
        }
```

Output below (first it starts as normal, then I change a file triggering a recompile, and you see several node_modules directories starting with @ choking, but all the other node_modules working fine).

```js
I20180512-13:46:28.692(2)? Meteor settings { public: {} }
I20180512-13:46:28.744(2)? Running shareDB server
=> Started your app.

=> App running at: http://localhost:3000/
{ Error: ENOENT: no such file or directory, unlink 'npm/node_modules/@babel'
    at Object.fs.unlinkSync (fs.js:1061:18)
    at Object.wrapper [as unlink] (/tools/fs/files.js:1666:20)
    at parts.forEach.part (/tools/isobuild/builder.js:159:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureDirectory (/tools/isobuild/builder.js:148:10)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:459:3)
    at optimisticReaddir.forEach.item (/tools/isobuild/builder.js:462:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:461:3)
    at Builder.copyNodeModulesDirectory (/tools/isobuild/builder.js:425:6)
    at /tools/isobuild/bundler.js:2307:17
    at Function._.each._.forEach (/Users/stian/.meteor/packages/meteor-tool/.1.7.0-rc.6.5p2mbn.25ps++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:2266:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2538:13)
    at /tools/isobuild/bundler.js:2662:30
    at Object.keys.forEach.name (/tools/isobuild/bundler.js:2797:9)
    at Array.forEach (<anonymous>)
    at /tools/isobuild/bundler.js:2787:26
    at /tools/isobuild/bundler.js:3128:22
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2961:31)
    at files.withCache (/tools/isobuild/bundler.js:2908:32)
    at Object.withCache (/tools/fs/files.js:1779:9)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2908:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: 'npm/node_modules/@babel' }
{ Error: ENOENT: no such file or directory, unlink 'npm/node_modules/@houshuang'
    at Object.fs.unlinkSync (fs.js:1061:18)
    at Object.wrapper [as unlink] (/tools/fs/files.js:1666:20)
    at parts.forEach.part (/tools/isobuild/builder.js:159:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureDirectory (/tools/isobuild/builder.js:148:10)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:459:3)
    at optimisticReaddir.forEach.item (/tools/isobuild/builder.js:462:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:461:3)
    at Builder.copyNodeModulesDirectory (/tools/isobuild/builder.js:425:6)
    at /tools/isobuild/bundler.js:2307:17
    at Function._.each._.forEach (/Users/stian/.meteor/packages/meteor-tool/.1.7.0-rc.6.5p2mbn.25ps++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:2266:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2538:13)
    at /tools/isobuild/bundler.js:2662:30
    at Object.keys.forEach.name (/tools/isobuild/bundler.js:2797:9)
    at Array.forEach (<anonymous>)
    at /tools/isobuild/bundler.js:2787:26
    at /tools/isobuild/bundler.js:3128:22
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2961:31)
    at files.withCache (/tools/isobuild/bundler.js:2908:32)
    at Object.withCache (/tools/fs/files.js:1779:9)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2908:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: 'npm/node_modules/@houshuang' }
{ Error: ENOENT: no such file or directory, unlink 'npm/node_modules/@material-ui'
    at Object.fs.unlinkSync (fs.js:1061:18)
    at Object.wrapper [as unlink] (/tools/fs/files.js:1666:20)
    at parts.forEach.part (/tools/isobuild/builder.js:159:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureDirectory (/tools/isobuild/builder.js:148:10)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:459:3)
    at optimisticReaddir.forEach.item (/tools/isobuild/builder.js:462:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:461:3)
    at Builder.copyNodeModulesDirectory (/tools/isobuild/builder.js:425:6)
    at /tools/isobuild/bundler.js:2307:17
    at Function._.each._.forEach (/Users/stian/.meteor/packages/meteor-tool/.1.7.0-rc.6.5p2mbn.25ps++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:2266:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2538:13)
    at /tools/isobuild/bundler.js:2662:30
    at Object.keys.forEach.name (/tools/isobuild/bundler.js:2797:9)
    at Array.forEach (<anonymous>)
    at /tools/isobuild/bundler.js:2787:26
    at /tools/isobuild/bundler.js:3128:22
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2961:31)
    at files.withCache (/tools/isobuild/bundler.js:2908:32)
    at Object.withCache (/tools/fs/files.js:1779:9)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2908:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: 'npm/node_modules/@material-ui' }
{ Error: ENOENT: no such file or directory, unlink 'npm/node_modules/@types'
    at Object.fs.unlinkSync (fs.js:1061:18)
    at Object.wrapper [as unlink] (/tools/fs/files.js:1666:20)
    at parts.forEach.part (/tools/isobuild/builder.js:159:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureDirectory (/tools/isobuild/builder.js:148:10)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:459:3)
    at optimisticReaddir.forEach.item (/tools/isobuild/builder.js:462:3)
    at Array.forEach (<anonymous>)
    at Builder._ensureAllNonPackageDirectories (/tools/isobuild/builder.js:461:3)
    at Builder.copyNodeModulesDirectory (/tools/isobuild/builder.js:425:6)
    at /tools/isobuild/bundler.js:2307:17
    at Function._.each._.forEach (/Users/stian/.meteor/packages/meteor-tool/.1.7.0-rc.6.5p2mbn.25ps++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
    at JsImage.write (/tools/isobuild/bundler.js:2266:7)
    at ServerTarget.write (/tools/isobuild/bundler.js:2538:13)
    at /tools/isobuild/bundler.js:2662:30
    at Object.keys.forEach.name (/tools/isobuild/bundler.js:2797:9)
    at Array.forEach (<anonymous>)
    at /tools/isobuild/bundler.js:2787:26
    at /tools/isobuild/bundler.js:3128:22
    at /tools/utils/buildmessage.js:271:13
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at exports.EnvironmentVariable.withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2961:31)
    at files.withCache (/tools/isobuild/bundler.js:2908:32)
    at Object.withCache (/tools/fs/files.js:1779:9)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2908:16)
    at Profile.run (/tools/runners/run-app.js:579:36)
    at Function.run (/tools/tool-env/profile.js:490:12)
    at bundleApp (/tools/runners/run-app.js:578:34)
    at AppRunner._runOnce (/tools/runners/run-app.js:622:35)
    at AppRunner._fiber (/tools/runners/run-app.js:880:28)
    at /tools/runners/run-app.js:408:12
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: 'npm/node_modules/@types' }
```

**Comment by benjamn** [comment]
I think the `try`-`catch` block is OK, but we should only hide the errors if the code is `ENOENT` (as it is in these examples).

**Comment by benjamn** [comment]
I believe this will be fixed if you run `meteor update --release 1.7-rc.7` and rebuild.

**Comment by houshuang** [comment]
I tried this, but had some crazy problems with rebuild times (I filed an issue, taking like 15 minutes for an initial build, instead of 30 seconds), so hard to test. I'm assuming this would fix it though, since exactly the same thing fixed it for me locally.
